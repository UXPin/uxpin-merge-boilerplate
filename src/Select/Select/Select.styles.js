import styled from "react-emotion";
import {css} from "emotion";
import * as colors from "../../styles/colors.json";
import * as borders from "../../styles/borders.json";
import * as shadow from "../../styles/shadows.json";
import * as typography from "../../styles/typography.json";
import * as size from "../../styles/sizes.json";

const SelectContainerStyles = styled("div")`
    width: ${props => props.stretched ? "100%" : "auto"};
    font-family: ${typography.fontFamily};
    font-weight: normal;
    font-size: ${typography.size.l}px;
    line-height: ${size.m * typography.lineHeightMultiplier}px;
    color: ${colors.gray.base};
`;

const SelectStyles = styled("div")`
    display: inline-flex;
    flex: 1 0 auto;
    align-items: center;
    justify-content: space-between;
    width: ${props => props.width === 'stretched' ? "100%" : props.width === 'auto' ? 'auto' : `${props.width}px`};
    min-width: 250px;
    height: ${size.xxl}px;
    margin-bottom: 3px;
    border-bottom: 1px solid ${colors.silver.dark20};

    button {
      margin-left: auto;  
    }
    svg {
      min-width: fit-content;
    }
`;

const SelectContent = styled("div")`
    overflow: hidden;
    white-space: nowrap;
    text-overflow: ellipsis;
    cursor: pointer;
`;

const MenuStyles = styled("ul")`
    position: absolute;
    display: flex;
    flex-direction: column;
    left: 0;
    width: ${props => props.width === 'stretched' ? "100%" : props.width === 'auto' ? 'auto' : `${props.width}px`};
    min-width: 250px;
    margin: 1px 0 0 6px;
    padding: ${size.s}px 0 ${size.s}px 0;
    background: white;
    box-shadow: ${shadow.box};
    list-style: none;
`

const ItemStyles = styled("li")`
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    justify-content: flex-start;
    position: relative;
    width: inherit;
    height: ${size.xxl}px;
    padding: 0 ${size.m}px;
    background-color: ${props => props.id === `${props.active}listItem` ? colors.silver.light1 : "white"};
    cursor: pointer;
    &:hover {
        background-color: ${colors.silver.light1};
    }
    span {
        max-width: 90%%;
        overflow: hidden;
        white-space: nowrap;
        text-align: center;
        text-overflow: ellipsis;
    }
`;

export { SelectStyles, MenuStyles, ItemStyles, SelectContainerStyles, SelectContent };