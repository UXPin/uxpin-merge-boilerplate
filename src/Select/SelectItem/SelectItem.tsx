import * as React from 'react';

export interface Props {
  children:string;
}

export default function SelectItem(props:Props):React.ReactNode {
  return (
    <span>
      {props.children}
    </span>
  );
}
