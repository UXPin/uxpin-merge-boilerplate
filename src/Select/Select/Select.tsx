import * as React from 'react';
import Button from '../../Button/Button';
import Icon from '../../Icon/Icon';
import { ItemStyles, MenuStyles, SelectContainerStyles, SelectContent, SelectStyles } from './Select.styles';

interface Props {
  children?:any;
  width?:'stretched' | 'auto' | number;
  placeholder?:string;
}

interface DefaultProps {
  width:'stretched';
  placeholder?:'Select...';
}

type defaultProps = Readonly<DefaultProps>;

export default class Select extends React.Component<Props, any> {
  private static defaultProps:defaultProps = {
    width: 'stretched',
    placeholder:'Select...'
  };
  constructor(props:Props) {
    super(props);
    this.state = {
      content:this.props.placeholder,
      open:false,
    };
  }

  public selectItem = (content:string, id:number) => {
    this.setState({
      content,
      open: false,
      selected:id,
    });
  }

  public toggleList = ():void => {
    this.setState(this.setState((prevState:any) => ({
      open: !prevState.open,
    })));
  }

  public restartSelect = (e:any):void => {
    e.stopPropagation();
    this.setState({ content:this.props.placeholder, open: false, selected: '' });
  }

  public render():any {
    const menuContent:React.ReactNode[] = this.props.children.length > 0 ? this.props.children : '';
    return (
      <SelectContainerStyles {...this.props}>
        <SelectStyles onClick={this.toggleList} {...this.props}>
            <SelectContent>
              {this.state.content}
            </SelectContent>
            {this.state.content !== 'Select...' ?
              <Button
                mode="minimal"
                type="error"
                size="s"
                stretched={false}
                icon={<Icon icon="ActionCloseSvg" size="xs" color="#444"/>}
                onClick={(e:any) => this.restartSelect(e)} />
            : ''}
            <Icon icon="ChevronSvg" color="#444"/>
        </SelectStyles>
        {this.state.open ?
        <MenuStyles {...this.props}>
            {menuContent.map((item:any, i:number) => {
              return <ItemStyles
                      id={`${i}listItem`}
                      active={this.state.selected}
                      onClick={() => this.selectItem(item.props.children, i)}
                      {...this.props}
                      >
                      {item}
                      </ItemStyles>;
            })}
        </MenuStyles> : ''}
      </SelectContainerStyles>
    );
  }
}
