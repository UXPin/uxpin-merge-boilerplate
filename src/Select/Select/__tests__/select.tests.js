import React from "react";
import * as enzyme from "enzyme";
import toJson from "enzyme-to-json";
import Select from "../Select.tsx";
import SelectItem from "../../SelectItem/SelectItem.tsx";

test("renders Select with width 300 correctly", () => {
  const select = (
    <Select width={300}>
      <SelectItem>Option 1</SelectItem>
      <SelectItem>Option 2</SelectItem>
      <SelectItem>Option 3</SelectItem>
      <SelectItem>Option 4</SelectItem>
      <SelectItem>Option 5</SelectItem>
      <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
    </Select>
  );

  expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
});

test("renders Select with width stretched correctly", () => {
    const select = (
      <Select width="stretched">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });

test("renders Select with width auto correctly", () => {
    const select = (
      <Select width="auto">
        <SelectItem>Option 1</SelectItem>
        <SelectItem>Option 2</SelectItem>
        <SelectItem>Option 3</SelectItem>
        <SelectItem>Option 4</SelectItem>
        <SelectItem>Option 5</SelectItem>
        <SelectItem>Option 5 Option 5 Option 5 Option 5 Option 5</SelectItem>
      </Select>
    );
  
    expect(toJson(enzyme.shallow(select))).toMatchSnapshot(`enzyme.shallow`);
  });