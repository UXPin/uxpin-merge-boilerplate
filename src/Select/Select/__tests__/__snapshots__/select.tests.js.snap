// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Select with width 300 correctly: enzyme.shallow 1`] = `
<Styled(div)
  placeholder="Select..."
  width={300}
>
  <Styled(div)
    onClick={[Function]}
    placeholder="Select..."
    width={300}
  >
    <Styled(div)>
      Select...
    </Styled(div)>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`renders Select with width auto correctly: enzyme.shallow 1`] = `
<Styled(div)
  placeholder="Select..."
  width="auto"
>
  <Styled(div)
    onClick={[Function]}
    placeholder="Select..."
    width="auto"
  >
    <Styled(div)>
      Select...
    </Styled(div)>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </Styled(div)>
</Styled(div)>
`;

exports[`renders Select with width stretched correctly: enzyme.shallow 1`] = `
<Styled(div)
  placeholder="Select..."
  width="stretched"
>
  <Styled(div)
    onClick={[Function]}
    placeholder="Select..."
    width="stretched"
  >
    <Styled(div)>
      Select...
    </Styled(div)>
    <Icon
      color="#444"
      icon="ChevronSvg"
    />
  </Styled(div)>
</Styled(div)>
`;
